### DISCUSSION TITLE
Add Symfony MakerBundle dependency

### DISCUSSION THREAD

**Discussion by tom-combet** [PrestaShop/PrestaShop#32246] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

As a developer, I want to use Symfony MakerBundle in PrestaShop so that I can generate redundant code super quickly.

### Describe the solution you'd like

Add `symfony/maker-bundle` package in core `composer.json`.

I know that the decision to add any package without using it directly in the core is tough.
However, this one is an official package from Symfony, and PrestaShop has now fully adopted this framework. Furthermore, it's not related to any dependency we don't use already.

### Alternatives you've considered

I already made [a tool](https://github.com/Kaudaj/prestashop-maker) based on the Symfony MakerBundle to generate code outside of PrestaShop. It then moves the generated code to a PrestaShop project afterwards.

It's definitely an ugly workaround to me and I'd prefer to turn this tool in a composer package. This would only be possible if symfony/maker-bundle is present in PrestaShop core.

### Additional context

Related to https://github.com/PrestaShop/PrestaShop/issues/25170.
I don't consider mine as a duplicate because I guess that the suggestion was creating a PrestaShop MakerBundle. I prefer by far extending Symfony MakerBundle with new makers for PrestaShop. If I'm wrong and it was the purpose of the issue, then feel free to close mine.

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by matks** [comment]
Hello @tom-combet I also like MakerBundle but consider the situation like this: if we add MakerBundle into PrestaShop project, it means when we build the next version 8.0.0 we'll ship MakerBundle with the project.

Let's say that 100,000 shops in the world upgrade from PS 1.7 to 8 . MakerBundle becomes part of these 100,000 shops. But... how many shops will find it useful? I'm sure there will be some hundred developers very happy to use it, but even if 5,000 developers it means we'll have 95,000 shops where MakerBundle is just present on the disk... and useless 😢 in their usecase.

When we ship PrestaShop we cannot ship it "with or without developer tools". It's one single ZIP.

**Comment by tom-combet** [comment]
Hi @matks,

Uh, I don't get it. Can't we just add it as a dev dependency ?

**Comment by matks** [comment]
@tom-combet Then I'm missing some informations 😄 

How do you develop? Do you use the sources of the project or do you use the ZIP archive? Do you run `composer install` in production?

**Comment by tom-combet** [comment]
@matks Oh no, I wouldn't use it in production of course.
Main use cases would be developing in PrestaShop core and developing third-party modules. But I get it now, third-party developers usually use ZIP archive so there is a problem... It would however remain very useful for the core.

**Comment by HanaRebaiQA** [comment]
Hello @tom-combet 

Thanks for your report.
@matks Following your answer, do you confirm that this feature will not be taken into consideration ?
If Yes, i'll close it.
Waiting for your feedback!

Thank you!

**Comment by matks** [comment]
> @matks Main use cases would be developing in PrestaShop core and developing third-party modules

What kind of features do you develop on the Core that could be boosted by MakerBundle? Honest question: I think we don't create new empty controllers, forms or tests ; most of the time we expand an already-existing file. This is why I ask.

Also you can add MakerBundle yourself (but don't commit it) 😄 actually on your computer you can do everything you do not need to insert MakerBundle in the upstream project to use it.

**Comment by tom-combet** [comment]
Adding Symfony MakerBundle would allow us to add other useful makers for grid, crud form, settings form, crud cqrs... in an extension package called PrestaShop Maker Bundle. Or directly in the core, like the commands.
I agree that existing ones will not be very useful, even if some of them could be nice like `make:command` or `make:entity`.

In fact, if the context is developing with the sources of the project, I can add it myself. But honestly, I think everyone could benefit from it to speed up Symfony migration and new features development.

By the way, why are the commands (`pr:mo`, `cache:clear`, ...) included in the PrestaShop archive if not to be used by developers? If we go in this way, we should remove them too, shouldn't we? 🤔

**Comment by matks** [comment]
> By the way, why are the commands (pr:mo, cache:clear, ...) included in the PrestaShop archive if not to be used by developers? If we go in this way, we should remove them too, shouldn't we? 🤔

For `cache:clear` we actually use it 😄 even in CI I believe

I don't know what is `pr:mo`.  But I guess it was simply inside Symfony from the beginning we did not add anything

**Comment by tom-combet** [comment]
Nope, [`pr:mo`](https://devdocs.prestashop.com/1.7/modules/testing/resources/#module-management-in-command-line) comes with PrestaShop. 😛 
And this is not the only one: see [src/PrestaShopBundle/Command](https://github.com/PrestaShop/PrestaShop/tree/develop/src/PrestaShopBundle/Command) directory.

**Comment by matks** [comment]
Ah, these Commands are meant to be used either in the project by contributors / maintainers or on live shops. We're sure they're being used. This is why they're in the project.

But I agree, we could remove the Commands dedicated to the project from the ZIP archive. Like [SecurityAnnotationLinterCommand.php](https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Command/SecurityAnnotationLinterCommand.php).

If you are motivated, you can submit a PR 👍 I'll approve it. The process that creates the ZIP is inside https://github.com/PrestaShop/PrestaShop/tree/develop/tools/build

**Comment by tom-combet** [comment]
> Ah, these Commands are meant to be used either in the project by contributors / maintainers or on live shops. We're sure they're being used. This is why they're in the project.

So for you, the commands would be more used than makers? This is highly debatable and very subjective to me.

By the way, you mean that merchants would run `prestashop:mail:generate`, `prestashop:list:commands-and-queries` and so on? Even if we talk about `pr:mo`, one of the most useful command, I'm not sure that many merchants use it...

For me, the target of these tools are developers. And by developers, I'm also talking about merchants who are learning how to develop in PrestaShop. I don't exclude them at all.
So, if target is developers, what makes commands more useful than makers? We can talk about real arguments then, for example the weight in the archive, the fact that makers will be useless in 1.6 shops... But the "We're sure they're being used." argument can't be verified easily.
Don't feel offended but I'm just asking: what would make you say that the makers would not be used? (once again, by developers I mean).

> But I agree, we could remove the Commands dedicated to the project from the ZIP archive. Like [SecurityAnnotationLinterCommand.php](https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Command/SecurityAnnotationLinterCommand.php).

Don't get me wrong, I'm not against commands in the archive. Rather the opposite. They are very useful and reinforce development and maintenance. This is also what motivates me to support the integration of maker-bundle.

> If you are motivated, you can submit a PR 👍 I'll approve it. The process that creates the ZIP is inside https://github.com/PrestaShop/PrestaShop/tree/develop/tools/build

If you're talking about removing commands from the archive, again, I'm against it, so I won't.

**Comment by matks** [comment]
> But the "We're sure they're being used." argument can't be verified easily.

For a faire number of these Commands I'm sure these Commands are used because... the maintainer team use it 😄  .

So basically we say "we can introduce MakerBundle because maybe some people will use it" while SecurityAnnotationLinterCommand I'm sure it's used because I use it myself 😋  .

Developer tools used by the maintainer team need to be included in the project
- because they need them to work on the project
- because the open source spirit is "if you use a tool, then you must make it available for the other project users" (we should not keep something "just for us" as much as possible)

So any developer tool needed and used by the maintainer team goes into the project.
In this topic we discuss about MakerBundle a tool that would not be used by the maintainer team (because today as far as I know we don't use it) but only by module developers.

It's hard you know 😅  I agree that PrestaShop can be seen as a development platform. But it's hard to decide what should go in the project and what should not. If we open too wide the door, we might add also [Rector](https://github.com/rectorphp/rector), [NelmioApiDocBundle](https://github.com/nelmio/NelmioApiDocBundle).

The scenario I want to avoid is: we accept your suggestion, we add MakerBundle, then only you @tom-combet use it, and we shipped MakerBundle into the 8.0.0 ZIP archive but nobody - excluding you - use it. That would be a failure.

If I was sure more than 70% of module developers and agency developers will use it I would accept 😁  but we have not this information.

**Comment by tom-combet** [comment]
> In this topic we discuss about MakerBundle a tool that would not be used by the maintainer team (because today as far as I know we don't use it) but only by module developers.

I already told that MakerBundle would not be very useful in itself, but with an extension made by PrestaShop.
By the way, we could imagine that we make the extension outside of PrestaShop before adding MakerBundle to the core, and add everything together once it's done. It would be maybe more logical.

Are you talking about MakerBundle itself or even with the future extension?

> It's hard you know 😅 I agree that PrestaShop can be seen as a development platform. But it's hard to decide what should go in the project and what should not. If we open too wide the door, we might add also [Rector](https://github.com/rectorphp/rector), [NelmioApiDocBundle](https://github.com/nelmio/NelmioApiDocBundle).

To me, it's not the same. Symfony IS already the framework used by PrestaShop. So the official tool MakerBundle is completely related. So, you would be against `DoctrineMigrationBundle` integration too, even if we use Doctrine?
I know it's hard to set limits, but here, this package is completely related to what we already have. So yea, question of point of view, hard debate I admit.

By the way, even if it's again not the same situation imo, I'm 100% pro-Rector as it would help us to migrate to PHP 8 more easily.
I don't really understand (but it's very personal here, and I don't expect you agree) why we should go without validated tools that speed up considerably the human tasks. Did the recent open-source controversies discouraged us to use them?
Well, an other debate here...

> The scenario I want to avoid is: we accept your suggestion, we add MakerBundle, then only you @tom-combet use it, and we shipped MakerBundle into the 8.0.0 ZIP archive but nobody - excluding you - use it. That would be a failure.

Yep, I agree that if I'm the only one who uses it, it's a complete failure of course.
I'm really not 100% sure about that, but I sadly can't prove the opposite easily (or maybe prove that 12 guys, the one who starred my repo, will use it haha)...

I hope I'm not wasting your time with this debate but I personally find it very interesting, even if we will surely not reach an agreement. It's always helpful to understand better the maintainers' mentality about this! 😃👍

**Comment by matks** [comment]
@tom-combet I'm happy to discuss this topic too 😄  and I agree: the discussion is valuable even if in the end we decide not to go this way

At this point I think some other people would be interesting to bring their opinion. Maybe @Matt75 @Quetzacoalt91 @kpodemski @PululuK @zuk3975 @Oksydan @Hlavtox ?

**Comment by PululuK** [comment]
I think there are two points here :

#### 1

- if a tool is needed only during the development phase it must be added as a dev dependency
- dev dependencies should be excluded during build

#### 2

On the other hand if a dev dependency should be added in the Core or not, this decision depends on the relevance and are real use in the core (it deserves a vote in this case).  Because adding a dependency also means maintaining it, and no one wants to maintain a tool they don't use.

**Comment by tom-combet** [comment]
Thanks for your opinion, @PululuK.
As I said, I think it could be used by third-party developers too, but I would already be happy if it could be integrated in dev dependencies!

Waiting the opinion of the other maintainers now... We had an interesting discussion with @kpodemski and @jolelievre about this on the slack server. Could you sum up your position there when you have some time please? 😃

**Comment by tom-combet** [comment]
@matks By the way, did you change your mind since https://github.com/PrestaShop/PrestaShop/issues/14086?

Because it completely changes the situation otherwise: third-party developers could use MakerBundle and its extension since they would use the sources and not the production ZIP in their dev environment.

**Comment by excitedbox** [comment]
I could see having easy admin bundle as being useful if someone wants to create custom admin interfaces, integrations of separate apps or components of their site that don't belong in the shopping cart.

It wouldn't be too hard to make it possible to add certain symfony bundles as needed but I also agree putting them in the production install for everyone just opens up a bunch of security issues. There could be a checkbox in the admin interface that deletes these extras if the user doesn't want them. If it is an opt in option you can avoid people keeping it by accident. Server resources are cheap and it seems foolish to argue about a few MB in a zip file if that was the only argument.

**Comment by matks** [comment]
Hi @tom-combet right now I don't feel confortable approving the idea of adding this 😞 

But I agree we need at some point to decide whether or not to include dev-targeted tools like this one. It's a tough decision but it's in our minds 👍

**Comment by tom-combet** [comment]
@excitedbox Nice idea to keep in mind, thanks!

@matks Hi, ok no pressure! 
Just wanted to know the current situation and it's great to know the direction PS is taking, thanks for that.

**Comment by Dreimus** [comment]
Adding the MakerBundle would be very usefull for symfony developers, as I could enhance better upgrade manipulation for Symfony Entities.

The possibility to add the Doctrine Migration would be another great feature for developers.
